src/script_opt/ZAM/Expr.cc

```diff
@@ -1100,23 +1100,84 @@ const ZAMStmt ZAMCompiler::ConstructSet(const NameExpr* n, const Expr* e) {
 
 const ZAMStmt ZAMCompiler::ConstructRecord(const NameExpr* n, const Expr* e) {
     auto rc = e->AsRecordConstructorExpr();
+    auto rt = e->GetType()->AsRecordType();
 
-    ZInstI z;
+    auto aux = InternalBuildVals(rc->Op().get());
 
-    if ( rc->Map() ) {
-        z = GenInst(OP_CONSTRUCT_KNOWN_RECORD_V, n);
-        z.aux = InternalBuildVals(rc->Op().get());
-        z.aux->map = *rc->Map();
-    }
-    else {
-        z = GenInst(OP_CONSTRUCT_RECORD_V, n);
-        z.aux = InternalBuildVals(rc->Op().get());
+    // Note that we set the vector to the full size of the record being
+    // constructed, *not* the size of any map (which could be smaller).
+    // This is because we want to provide the vector directly to the
+    // constructor.
+    aux->zvec.resize(rt->NumFields());
+
+    if ( pfs->HasSideEffects(SideEffectsOp::CONSTRUCTION, e->GetType()) )
+        aux->can_change_non_locals = true;
+
+    ZOp op;
+
+    auto map = rc->Map();
```

Comment:
This runs just once during compilation? Probably doesn't matter at all, but I believe there's an extra copy of the whole map here.

 ```suggestion
     auto& map = rc->Map();
```

(and possibly const, too)

Reply:
Whoops thanks for catching that.